core/src/worker/activities.rs

```diff
@@ -509,6 +527,53 @@ where
                                     tt,
                                     ActivityCancelReason::WorkerShutdown,
                                 ));
+                            } else {
+                                // Fire off task to keep track of local timeouts. We do this so that
+                                // activities can still get cleaned up even if the user isn't
+                                // heartbeating.
+                                let local_timeout_buffer = self.local_timeout_buffer;
+                                let sched = task
+                                    .resp
+                                    .schedule_to_close_timeout
+                                    .clone()
+                                    .and_then(|to| {
+                                        // Must filter out zero before any subtraction math here
+                                        if to.seconds == 0 && to.nanos == 0 {
+                                            None
+                                        } else {
+                                            Some(to)
+                                        }
+                                    })
+                                    .zip(task.resp.scheduled_time.clone())
+                                    .and_then(|(to, st)| {
+                                        let now = SystemTime::now();
+                                        let sched_time = SystemTime::try_from(st).unwrap_or(now);
+                                        let already_elapsed =
+                                            now.duration_since(sched_time).unwrap_or_default();
+                                        Duration::try_from(to).map(|to| to - already_elapsed).ok()
+                                    });
+                                let timeout_at = [
+                                    task.resp.heartbeat_timeout.clone(),
+                                    task.resp.start_to_close_timeout.clone(),
+                                ]
+                                .into_iter()
+                                .flatten()
+                                .map(Duration::try_from)
+                                .filter_map(Result::ok)
+                                .chain(sched)
+                                .filter(|d| !d.is_zero())
+                                .min();
+                                if let Some(timeout_at) = timeout_at {
+                                    let cancel_tx = cancels_tx.clone();
+                                    tokio::task::spawn(async move {
```

Comment:
It would be nice not to leave dangling tasks around and instead merge this work into some stream that can be aborted when no longer needed or at least have a reference to this task so it can be canceled when the activity resolves.

Reply:
I generally agree but there is no real need to track these tasks at all - nothing can go wrong in them. That said cancelling when the activity resolves would be useful for sure. I'll find a way to add that.